core-keys: Map `M-m` to mode-map when available

justbur opened the pull request:
This is how evil-leader gets `SPC m` to work for mode specific bindings,
and we need to mirror this step for `M-m`

justbur commented:
@TheBB This is still not totally correct. The mode-maps need to be mapped to the leader keys everytime they're available. evil-leader does this for SPC, but we need to replicate this behavior for the emacs leader key to get it to work properly. Before this change, the emacs leader key was only mapped to `evil-leader--default-map` which makes the mode-specific bindings unavailable (`M-m m`).

justbur commented:
@TheBB @syl20bnr This turned into a bigger rewrite than I intended, but this is the right way to do this. Please test it yourselves, but for comparison look at the mode hook in the evil-leader package. Everytime it fires evil-leader looks up the mode-map and binds it to the leader key if it exists. We need to do the same thing for `M-m` or `M-m m` won't work.

justbur commented:
@syl20bnr I see you put this on the 0.105 milestone, but my idea was to use #3019 as the next step. This PR was meant to keep the current use of evil-leader but patch up the inconsistencies. Like `M-m m` never working in the current develop branch. Or the issue pointed out in #3036.

syl20bnr commented:
Fine with me, I'll review it tonight with the remaining PRs.

justbur commented:
Ok, I know it seems like a big change, but I'm following the [evil-leader implementation](https://github.com/cofi/evil-leader/blob/master/evil-leader.el#L123-L161) almost exactly.

syl20bnr commented:
I hope you can reproduce it: I just cherry-picked it and restarted emacs, SPC leader was not defined.

syl20bnr commented:
I use Vim style.

justbur commented:
That's strange. The function needs to be called once. It was just not getting triggered on the initial load

justbur commented:
I updated the pr

justbur commented:
There's something I'm still not getting about use-package here. According to the docs for use-package the :init code runs before the package is loaded, but the only way the `global-evil-leader-mode` call can work here is if the package is autoloaded through package.el when the init forms are run, right? So doesn't that mean that when it comes time for use-package to load evil-leader, it doesn't do anything?

justbur commented:
I didn't realize the new function wasn't being called for the very first buffer. I don't completely understand why, but moving everything to the init part of use-package resolves the problem for me.

justbur commented:
Sorry for the spam, but I'm finally done tweaking this. The leader not being available in the spacemacs buffer (or the messages buffer for that matter) seems to be a deficiency of evil-leader itself https://github.com/cofi/evil-leader/issues/10. 

The old code here was inadvertently fixing that problem by binding the leader in motion state from the start. Since I switched to following the evil-leader way of doing things, I accidentally introduced that problem. The code is designed now so that it follows exactly what evil-leader does. I added a loop to force evil-leader mode in the existing buffers on startup.

syl20bnr commented:
Excellent :+1: 
Thank you !
Cherry-picked into develop branch, you can safely delete your branch.

syl20bnr commented:
New regression, in an org buffer, `SPC b` is undefined. Only a few key bindings are missing.

justbur commented:
@syl20bnr I can't reproduce that problem. What state were you in?

justbur commented:
Sorry I was on the wrong commit. I see it now

TheBB commented:
I can reproduce it. You're aware that this PR was reverted right?

justbur commented:
@syl20bnr @TheBB Why are we doing [this](https://github.com/syl20bnr/spacemacs/blob/develop/layers/org/packages.el#L46-L50)? That's the problem here

Why would evil-org mode bind to the leader key?

syl20bnr commented:
It works but the prefix name are reset in which-key.

Thank you ! :+1: 
Cherry-picked intro develop branch, you can safely delete your branch.

justbur commented:
Those org ones or all of them?
On Sat, Sep 19, 2015 at 1:46 AM Sylvain Benner <notifications@github.com>
wrote:

> It works but the prefix name are reset in which-key.
>
> Thank you ! [image: :+1:]
> Cherry-picked intro develop branch, you can safely delete your branch.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/syl20bnr/spacemacs/pull/3031#issuecomment-141623340>.
>

syl20bnr commented:
The org ones only, also :point_up: [19 septembre 2015 00:09](https://gitter.im/syl20bnr/spacemacs?at=55fce00faef03edd3ed14b86)

justbur commented:
I'm not seeing any problem with the prefix names in org mode. 

Can you reproduce that comment in erc? I don't use that layer
